src/result_processor.c

```diff
@@ -845,3 +846,203 @@ ResultProcessor *RPCounter_New() {
+  SearchResult *curr_res = FixedSizeBlocksManager_getNextElement(&RPBuffer->resultsIterator);
+  
+  // iterate the buffer.
+  while(curr_res) {
+    // Skip invalid results
+    if (isResultValid(curr_res)) {
+      return ReturnResult(curr_res, result_output);
+    }
+
+    // If the result is invalid discard it.
+    SearchResult_Clear(curr_res);
+    curr_res = FixedSizeBlocksManager_getNextElement(&RPBuffer->resultsIterator);
+  }
+
+  return RS_RESULT_EOF;
+}
+
+int rpbufferNext_bufferDocs(ResultProcessor *rp, SearchResult *res) {
+  RPBufferAndLocker *RPBuffer = (RPBufferAndLocker *)rp;
+
+  // Get the current index version so we can check later if any updates accured
+  // after we finish to buffer the results
+  RedisSearchCtx *RSctx = RPBuffer->base.parent->sctx;
+  double currentIndexVersion = IndexSpec_GetVersion(RSctx->spec);
+
+  FixedSizeBlocksManager *buffer = &RPBuffer->bufferBlocks;
+
+  // Allocate the first results block.
+  FixedSizeBlocksManager_init(buffer, sizeof(SearchResult),
+                                      RPBuffer->BlockSize);
+
+  // Keep fetching results from the upstream result processor until EOF is reached
+  while (1) {
+
+    // Get a space for a new result.
+    SearchResult *resToBuffer = FixedSizeBlocksManager_getEmptyElement(buffer);
+
+    // Get the next result and save it in the buffer
+    int result_status = rp->upstream->Next(rp->upstream, resToBuffer);
+
+    // If upstream has finished stop buffering results or something went wrong
+    if (result_status != RS_RESULT_OK) {
+      if (result_status == RS_RESULT_EOF ||
+          (result_status == RS_RESULT_TIMEDOUT &&
+           RSGlobalConfig.timeoutPolicy == TimeoutPolicy_Return)) {
+        // Break the loop and continue to the next step.
+        if (!FixedSizeBlocksManager_isEmpty(buffer)) {
+          break;
+        }
+      }  // else, we got no results or something went wrong, we can go downstream.
+      return result_status;
+    }
+  }
+
+  // Now we have the data of all documents that pass the query filters,
+  // let's lock the GIL to provide safe access to Redis keyspace 
+  
+  // initialize next function to the defualt value
+  rp->Next = rpbufferNext_Yield;
+
+  int lockGIL = RedisModule_ThreadSafeContextTryLock(RSctx->redisCtx);
+
+  // TryLock returns REDISMODULE_ERR if the GIL was already locked
+  if (lockGIL == REDISMODULE_ERR) {
+
+    // unlockspec to avoid deadlocks
+    RedisSearchCtx_UnlockSpec(RSctx);
+
+    // Lock GIL to gurentee safe access to Redis keyspace
+    RedisModule_ThreadSafeContextLock(RSctx->redisCtx);
+
+    // If the spec has been changed since we released the spec lock,
+    // we need to validate every buffered result
+    if (currentIndexVersion != IndexSpec_GetVersion(RSctx->spec)) {
+      rp->Next = rpbufferNext_ValidateAndYield;
+    }
+  }
+
+  // If we managed to lock the GIL it means that
+  // nobody held it, so we can't have a deadlock.
```

Comment:
you still need to unlock the spec for this comment to be valid

Reply:
So checking if Redis GIL is locked is redundant actually